_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,133 @@
+        }
+      }
+    }
+    ```
+   `BufferedLogger` is an [active object](https://en.wikipedia.org/wiki/Active_object). Method `writeStatementsInBackground` itself has control flow responsibility.
+ 
+   In a distributed system, examples of services with this type of responsibility could be a proxy, a load balancer, or a service transparently caching responses from or buffering requests to some other service.
+
+#### How small or large a responsibility should be?
+I hope the examples above give some more grounded sense of what a responsibility of a method, class, module, or service could be. However, they still provide no actionable guidance on how finely you should chop responsibilities between the components of your system. For example:
+
+ - Should conversion from XML to JSON be a responsibility of a single method (or a class), or it should be split between two methods: one translates XML into a tree, and another serializes a tree into JSON? Or these should be separate methods belonging to a single class?
+ - Should individual types of interactions with an external service (such as different RPC operations) be responsibilities of separate classes,  or they should all belong to a single class? Or, perhaps, interactions should be grouped, such as read operations going to one class and write operations going to another?
+ - How small parts of control flow should be encapsulated in individual methods?
+ - What kinds of responsibilities should be split between different (micro)services?
+
+Uncle Bob Martin (who first proposed Single Responsibility Principle) suggests that components should be broken down until each one has only one *reason to change*. However,  to me, this criterion still doesn't feel very instructive. Consider the `processTransaction` method above. There may be many reasons to change it:
+ - Increment counters of normal and fraud transactions to gather statistics.
+ - Enrich or reformat the logging statement.
+ - Wrap sending an alert into error-handling `try-catch` and log a failure to send an alert.
+ 
+Does this mean that `processTransaction` method is too large and we should split it further into smaller methods? According to Uncle Bob, [we probably should](https://sites.google.com/site/unclebobconsultingllc/one-thing-extract-till-you-drop), but many other people may think that `processTransaction` is already small enough.
+
+Let's return to the purpose of using Single Responsibility Principle. Obviously, it's to improve the overall quality of the codebase and of its production behavior (Carlo Pescio calls these [artifact space and run-time space](http://www.carlopescio.com/2010/06/notes-on-software-design-chapter-6.html), respectively).
+
+So, what will utimately help us to apply Single Responsibility Principle effectively is **more clarity into how using the principle affects the quality of the code and the running application**. The optimal scope of the responsibility for a component highly depends on the context:
```

Comment:
The sentence "more clarity into how using the principle affects the quality of the code and the running application" takes a lot of my cognitive resources to process :). When you go over your text again correcting things, please try to think of ways how to make sentences like this one easier, perhaps by breaking them up into two sentences.

Reply:
Rephrased to "So, what will ultimately help us to apply the Single Responsibility Principle effectively is **making clearer for ourselves how SRP affects the quality of the code and the running application**."